### PR TITLE
Feature/apl 648 update shard requirements

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
Added code prepared for adding shard steps.

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/Blockchain.java [hunk 1]
```diff
@@ -64,6 +65,8 @@
 
     void commit(Block block);
 
+    Long getBlockCount(TransactionalDataSource dataSource, int from, int to);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/Blockchain.java]: I think, that blockchain should not take datasource as parameter. Maybe specify datasource by id?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TransactionDaoImpl.java [hunk 2]
```diff
@@ -324,6 +324,28 @@ public int getTransactionCount() {
...
     }
 
+    @Override
+    public Long getTransactionCount(TransactionalDataSource dataSource, int from, int to) {
+        if (dataSource == null) {
+            // select from main db
+            dataSource = databaseManager.getDataSource();
+        }
+        try (Connection con = dataSource.getConnection()) {
+            PreparedStatement pstmt = con.prepareStatement("SELECT count(*) as transactionCount FROM transaction WHERE height >= ? AND height < ?");
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TransactionDaoImpl.java]: Maybe simplify to count(*) without `as transactionCount` and then `getLong(1)`?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/model/Shard.java [hunk 3]
```diff
@@ -132,8 +149,33 @@ public ShardBuilder shardHeight(Integer shardHeight) {
...
+        if (shardHash == null) sb.append("null");
+        else {
+            sb.append('[').append(shardHash).append(']');
+        }
+        sb.append(", shardState=").append(shardState);
+        sb.append(", shardHeight=").append(shardHeight);
+        sb.append(", zipHashCrc=");
+        if (zipHashCrc == null) sb.append("null");
+        else {
+            sb.append('[').append(zipHashCrc).append(']');
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/model/Shard.java]: Covert bytes to hexString pls

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java [hunk 4]
```diff
@@ -267,52 +284,57 @@ public MigrateState relinkDataToSnapshotBlock(CommandParamInfo paramInfo) {
...
-        log.debug("RELINK Processed table(s)=[{}] in {} secs", paramInfo.getTableNameList().size(), (System.currentTimeMillis() - startAllTables) / 1000);
+        log.debug("RELINK Processed table(s)=[{}] in {} secs", paramInfo.getTableNameList().size(), (System.currentTimeMillis() - startAllTables)/1000);
         return state;
     }
+*/
 
     private void processOneTableByHelper(CommandParamInfo paramInfo, Connection sourceConnect,
                                          String tableName, long start,
-                                         BatchedPaginationOperation batchedPaginationOperation) throws Exception {
+                                         Optional<BatchedPaginationOperation> paginationOperationHelper) throws Exception {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java]: Do not use optional as parameter: https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments/31923105

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/Blockchain.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/Blockchain.java]: We don't know the DataSource by Id beforehand, only in test it's 'kind of fixed'.
Here I wanted decrease injected dependencies in unit test for checking rows in main or shard db.
I can remove those methods on Blockchain service, but Inject only BlockDao + TransactionDao into unit test.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java (hunk 4 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/model/Shard.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/dao/model/Shard.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TransactionDaoImpl.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/TransactionDaoImpl.java]: Done

**Pull request merged**